libs2eplugins/src/s2e/Plugins/Core/BaseInstructions.cpp

```diff
@@ -723,41 +723,22 @@ void BaseInstructions::getConstraintsCountForExpression(S2EExecutionState *state
  */
 void BaseInstructions::forkCount(S2EExecutionState *state) {
     target_ulong count;
-    target_ulong nameptr;
 
     state->jumpToSymbolicCpp();
 
     state->regs()->read(CPU_OFFSET(regs[R_EAX]), &count, sizeof count);
-    state->regs()->read(CPU_OFFSET(regs[R_ECX]), &nameptr, sizeof nameptr);
-
-    std::string name;
-
-    if (!state->mem()->readString(nameptr, name)) {
-        getWarningsStream(state) << "Could not read string at address " << hexval(nameptr) << "\n";
-
-        state->regs()->write<target_ulong>(CPU_OFFSET(regs[R_EAX]), -1);
-        return;
-    }
-
-    klee::ref<klee::Expr> var = state->createSymbolicValue<uint32_t>(name, 0);
-
-    state->regs()->write(CPU_OFFSET(regs[R_EAX]), var);
     state->regs()->write<target_ulong>(CPU_OFFSET(eip), state->regs()->getPc() + 10);
 
-    getDebugStream(state) << "s2e_fork: will fork " << count << " times with variable " << var << "\n";
-
-    for (unsigned i = 1; i < count; ++i) {
-        klee::ref<klee::Expr> val = klee::ConstantExpr::create(i, var->getWidth());
-        klee::ref<klee::Expr> cond = klee::NeExpr::create(var, val);
+    getDebugStream(state) << "s2e_fork: will fork " << count << " time(s)\n";
 
-        klee::Executor::StatePair sp = s2e()->getExecutor()->forkCondition(state, cond, true);
+    for (unsigned i = 0; i < count; ++i) {
+        S2EExecutor::StatePair sp = s2e()->getExecutor()->fork(*state);
         assert(sp.first == state);
         assert(sp.second && sp.second != sp.first);
-    }
 
-    klee::ref<klee::Expr> cond = klee::EqExpr::create(var, klee::ConstantExpr::create(0, var->getWidth()));
-    if (!state->addConstraint(cond)) {
-        s2e()->getExecutor()->terminateState(*state, "Could not add condition");
+        S2EExecutionState *forkedState = static_cast<S2EExecutionState *>(sp.second);
+        klee::ref<klee::Expr> index = klee::ConstantExpr::create(i, klee::Expr::Int32);
+        forkedState->regs()->write(CPU_OFFSET(regs[R_EAX]), index);
     }
```

Comment:
I tried to write concrete `index` into RAX of the forked state, but it shows me this assertion failure:

```
5 [State 0] BaseInstructions: s2e_fork: will fork 1 times
5 [State 0] Forking state 0 at pc = 0x55a7aebed33d at pagedir = 0xdabe000
    state 0
    state 1
qemu-system-x86_64: /home/aesophor/s2e/source/s2e/libs2ecore/src/S2EExecutionStateRegisters.cpp:212: 
void s2e::S2EExecutionStateRegisters::writeSymbolicRegion(unsigned int, const void *, unsigned int):
Assertion `*m_active' failed.
./launch-s2e.sh: line 97: 146633 Aborted                 
(core dumped) LD_PRELOAD=$LIBS2E $QEMU $QEMU_DRIVE -k en-us
$GRAPHICS -monitor null -m $QEMU_MEMORY -enable-kvm -serial file:serial.txt
$QEMU_EXTRA_FLAGS -loadvm $QEMU_SNAPSHOT $*
```

Previously, `forkCount()` writes the return value by creating a symbolic variable for the current state (with value being 0), and write that symbolic variable into RAX.

I'm not sure if I should create that symbolic variable, so instead I just create a `klee::ConstantExpr`, write it into RAX of the forked state, and let `s2e_fork()` concretize the return value.

Reply:
I see, I forgot about this limitation. Let's then postpone this for later and keep `void s2e_fork()`.

Reply:
Sure! I've reverted it back.